# Проект по анализу данных

---

**Тема проекта:** анализ восприятия нейросетей на примере контента в YouTube.

**Исследовательский вопрос**: каким образом пользователи ютуб регируют на контент связанный с  нейросетями?

**Задачи:**
1. Просмотреть связана ли вовлеченность в контент (видео) с категорией видео (первая переменная будет создана на основании переменных из базы; вторая переменная взята из базы в исходном виде).
2. Выяснить, связана ли вовлеченность в контент (видео) с годом выпуска видео (обе переменные будут созданы на основании переменных в базе).


In [ ]:
# № 0. устнановим и импортируем нужные нам  модули/библиотеки
!pip install randan
#!pip install plotly

import googleapiclient.discovery as api # установленный модуль для применения API (только для неавторизованных)
import pandas
import numpy
import math
from randan.descriptive_statistics import ScaleStatistics # количественные переменные
from randan.descriptive_statistics import NominalStatistics # номинальные переменные
from randan.bivariate_association import Correlation # корреляция
from randan.bivariate_association import Crosstab # Хи-квадрат
import plotly.express as px
import random, requests, time # генерация случайного числа, обращения по интересующему URL-адресу и "засыпания" исполнения соответственно

In [18]:
# № 1 авторизируемся по ключу

# добавим ключ для авторизации

API_KEY = "тут ваш ключ апи"
# указали название сервиса
api_service_name = "youtube"

# версия API
api_version = "v3"

# все необходимые данные для авторизации
youtube = api.build(api_service_name, api_version, developerKey = API_KEY)

In [19]:
# № 2 формирование запроса

# запрос
q = "нейросеть"

# источник
type = "video"

In [20]:
# № 3 подача запроса
request = youtube.search().list(
    part = "snippet"             # snippet отвечает за выдачу набора параметров видео и каналов
    , maxResults = 50
    , q = q
    , type = type)

# сохраняем все собранные данные по запросу в объект response
response = request.execute()

#  это датафрейм, в котором будет сохранена вся информация о первых 50 видео
df_ne = pandas.json_normalize(response["items"])

df_ne

kind                         etag        id.kind  \
0   youtube#searchResult  TRX0qxSg_4NLd2WOzb8UEWddKrA  youtube#video   
1   youtube#searchResult  QEeuxSXO7_Lzkme5dzGODj3KR4I  youtube#video   
2   youtube#searchResult  aH_mQ7JlBAOBwVR0ACA8FKybqHg  youtube#video   
3   youtube#searchResult  X7Ho0HNx39H7Dg_wdkGbRZSg2jE  youtube#video   
4   youtube#searchResult  Rhdqx9OKctwmfp-s48knyxXC3p8  youtube#video   
5   youtube#searchResult  YRtsnpWq6u0KME3WfV_q8Lqr4wc  youtube#video   
6   youtube#searchResult  8W78t5a4HRvFr46OpMS_ksKTaNc  youtube#video   
7   youtube#searchResult  KIsO5ktn7bTaTq6CjtEqbUJZoqw  youtube#video   
8   youtube#searchResult  fmcRc4NCajSmVAIQDkUPPP-l670  youtube#video   
9   youtube#searchResult  CtMHab1S8sZyrm-bwhxP4vDQ_Qc  youtube#video   
10  youtube#searchResult  CQEucFBQ3LYmtzbW5KHd7NeHSqs  youtube#video   
11  youtube#searchResult  pPAMcIiTatlXMCDVZmQaYggqSkQ  youtube#video   
12  youtube#searchResult  SOZVF1U8KibdbW4VUzopnQ48PUw  youtube#video   
13  youtube#searchResult  niu5cqzxFZ_CyZm4sMNqJcEYMFw  youtube#video   
14  youtube#searchResult  ZBYXZSqOtr8f5vHWkEm9ZliElCc  youtube#video   
15  youtube#searchResult  XraszV1ws5-SibzSvamHlDS7wJE  youtube#video   
16  youtube#searchResult  afPyOJ_oHX-mA1gTF3IPHC3Q1o8  youtube#video   
17  youtube#searchResult  aWisuTZwSASaxi4VGpggRnQQMiU  youtube#video   
18  youtube#searchResult  8zi32FzjjVosdqAIC0yUUajtHrw  youtube#video   
19  youtube#searchResult  BYIGNXheA5jr3IN9uwGRkZUh29U  youtube#video   
20  youtube#searchResult  xKlAN4FY4ZHe8ay4oQi83LB3FIU  youtube#video   
21  youtube#searchResult  AUsJcOL77vSFQZPfxaur0tsddqQ  youtube#video   
22  youtube#searchResult  ZEn6fl30Cqc553qeiACY14Ohuj4  youtube#video   
23  youtube#searchResult  dxv4h9ahuptyF44fL7y24uIK8n0  youtube#video   
24  youtube#searchResult  ZnMEpMvK-rNXYhWgIZBqXY_Vsvc  youtube#video   
25  youtube#searchResult  tS_svw-oDFj7h2HN3jYimD6tuDs  youtube#video   
26  youtube#searchResult  Z88vDq-xUdZW47RxLprptbpTdBg  youtube#video   
27  youtube#searchResult  _6IBMNxjq01KPN1Bn5plVklbRbg  youtube#video   
28  youtube#searchResult  qbYqxMNQOBFCZQ8vwBMSj2I_dK0  youtube#video   
29  youtube#searchResult  vdtxCFL-A5T9e4fo0hWQeQ3KOdY  youtube#video   
30  youtube#searchResult  TgyVIN_zjNNhOq0f8GpPREIjPMs  youtube#video   
31  youtube#searchResult  kYAjD7PXiqynJ8KKio3pkaK7Azc  youtube#video   
32  youtube#searchResult  4FlaR4Rf4TPtoRoUkIwm77YK2YE  youtube#video   
33  youtube#searchResult  7C71UoICjMRrLb_S_2Q4jkZbB-E  youtube#video   
34  youtube#searchResult  6i3oSHbQo7fSOrPuxaJQULyY8ds  youtube#video   
35  youtube#searchResult  CuCPNdZACrPH2GBMGCcT0kHmVr4  youtube#video   
36  youtube#searchResult  XRtb-49q1YZhrCCsXyhNVOmQkzs  youtube#video   
37  youtube#searchResult  CbQU83Nqkb30PX4GJQ2sjX886kI  youtube#video   
38  youtube#searchResult  8g8CxiVpihXQz_6U10BgYwlIDLs  youtube#video   
39  youtube#searchResult  f6NuAczWyx-VfhgIyaj776dC7Og  youtube#video   
40  youtube#searchResult  jgRcpfZbUqDSct3LeeBiQco2SlA  youtube#video   
41  youtube#searchResult  tmMoGbsS_lvngspnqSv07rak4u4  youtube#video   
42  youtube#searchResult  8FUU7_rDiDn4MnP8AwObOYWn9OI  youtube#video   
43  youtube#searchResult  k_UjnNVfnRa3Xvhiv_rEoRC6qFk  youtube#video   
44  youtube#searchResult  Sgf-fu1JKaoveOzy-S7eURyJFdU  youtube#video   
45  youtube#searchResult  gEQdjbmNfnE0NDefP0pdy-4FruU  youtube#video   
46  youtube#searchResult  RsCgaOMAWKpb9ILeoI66LS2CbrE  youtube#video   
47  youtube#searchResult  SpYc_oWWzLX1K-LgAVz6b88ic9I  youtube#video   
48  youtube#searchResult  RrzjDgIzhxmy37s4XUK-gNmywgw  youtube#video   
49  youtube#searchResult  -Wmfbd8dCc9douGtUOLtvcQWPlc  youtube#video   

     id.videoId   snippet.publishedAt         snippet.channelId  \
0   2wEAXtfNKKk  2023-05-14T15:00:24Z  UCr1an43NIMXKgZe18zZZd4Q   
1   Y6pgNMQlbMA  2023-02-01T12:34:43Z  UCuaYG7fdQ-4myL_CVtvwNHQ   
2   WIttZGQEdas  2020-10-08T20:04:08Z  UCY03gpyR__MuJtBpoSyIGnw   
3   CMWnaPW04qc  2023-06-14T12:23:26Z  UCuaYG7fdQ-4myL_CVtvwNHQ   
4   csIxOuVN

In [21]:
# № 4 смотрим количество существующих видео на момент подачи запроса
display(response["pageInfo"]["totalResults"])

341486

In [22]:
# № 5 обозначение переменной для тех видео, что ищутся по другим сортировкам
order_list = ["date", "rating", "relevance", "title", "videoCount", "viewCount"]
order_list

['date', 'rating', 'relevance', 'title', 'videoCount', 'viewCount']

In [23]:
#  6  запроса с использованием цикла для сбора данных со всех страниц с выдачей и аргумента order для увеличения выдачи
i = 0

# lанный цикл будет работать до тех пор, пока не пройдется по всем значениям аргумента order
for order in order_list:

    if len(df_ne.drop_duplicates("id.videoId")) < response["pageInfo"]["totalResults"]:
        request = youtube.search().list(
            part = "snippet"
            , maxResults = 50
            , q = q
            , type = type
            , order = order)

        response = request.execute()

        print(f'Итерация №{i}, "order" {order}, "items" {len(response["items"])}')
        i += 1

        df = pandas.pandas.json_normalize(response["items"]) # датафрейм, в котором обработанные данные
        df_ne = pandas.concat([df, df_ne]) # датафрейм, в котором новые данные (df) с каждой страницы (итерации)

        # данный цикл будет работать до тех пор, пока на текущей странице указан токен следующей страницы (nextPageToken)
        while ("nextPageToken" in response.keys())\
        & (len(df_ne.drop_duplicates("id.videoId")) < response["pageInfo"]["totalResults"])\
        & (len(response["items"]) > 0):
              request = youtube.search().list(
                  part = "snippet"
                  , maxResults = 50
                  , q = q
                  , type = type
                  , pageToken = response["nextPageToken"]
                  , order = order)

              response = request.execute()

              print(f'Итерация №{i}, "order" {order}, "items" {len(response["items"])}')
              i += 1
              df_additional = pandas.json_normalize(response["items"])
              df_ne = pandas.concat([df_ne, df_additional])

df_ne

Итерация №0, "order" date, "items" 50
Итерация №1, "order" date, "items" 50
Итерация №2, "order" date, "items" 50
Итерация №3, "order" date, "items" 50
Итерация №4, "order" date, "items" 50
Итерация №5, "order" date, "items" 50
Итерация №6, "order" date, "items" 50
Итерация №7, "order" date, "items" 50
Итерация №8, "order" date, "items" 50
Итерация №9, "order" date, "items" 50
Итерация №10, "order" date, "items" 50
Итерация №11, "order" date, "items" 46
Итерация №12, "order" rating, "items" 50
Итерация №13, "order" rating, "items" 50
Итерация №14, "order" rating, "items" 50
Итерация №15, "order" rating, "items" 50
Итерация №16, "order" rating, "items" 50
Итерация №17, "order" rating, "items" 50
Итерация №18, "order" rating, "items" 50
Итерация №19, "order" rating, "items" 50
Итерация №20, "order" rating, "items" 50
Итерация №21, "order" rating, "items" 50
Итерация №22, "order" rating, "items" 50
Итерация №23, "order" rating, "items" 39
Итерация №24, "order" relevance, "items" 50
Итерац

kind                         etag        id.kind  \
0   youtube#searchResult  dxv4h9ahuptyF44fL7y24uIK8n0  youtube#video   
1   youtube#searchResult  MwUFhIFJ2sm_-IOduhHKJVw1A2s  youtube#video   
2   youtube#searchResult  C2Fk0U47OES4qJIEUmzvs9pVRfc  youtube#video   
3   youtube#searchResult  BQcCdIuFNFleZ_sQGot_Zq1YNgQ  youtube#video   
4   youtube#searchResult  KR12pERv9PDvaiayuGB_tApP4ZU  youtube#video   
..                   ...                          ...            ...   
30  youtube#searchResult  zr7h1hocrhSiXF85ULZgxmuP3-I  youtube#video   
31  youtube#searchResult  MDVtMD7Y6OrDXoaR4Wnm8z-xGLY  youtube#video   
32  youtube#searchResult  9U4fvWFkyJ1uoRZLAh4JexPJGuE  youtube#video   
33  youtube#searchResult  7UiKws_Bk8zTWtdZXaEiXffXZbo  youtube#video   
34  youtube#searchResult  IX4G67GeG77ybEgzSDRZl2KcY-Q  youtube#video   

     id.videoId   snippet.publishedAt         snippet.channelId  \
0   3OxtnK1m5Sw  2022-10-20T22:44:44Z  UCm7bpbe7_H_47ef4dJH0SNw   
1   h6QZtd-Efko  2023-05-04T09:59:13Z  UCf2ZCvUQRwdkFUfebyy0Xag   
2   HaLZtHN83JE  2023-01-26T13:00:15Z  UCRd8_F4-SynaG4sE5lnXv_g   
3   gkDrVYInSaQ  2023-02-10T14:00:24Z  UCUEQBzSZx65-FcVwce8HYNQ   
4   P8GXJ7AWqIA  2022-12-05T07:04:35Z  UCkF05U4kMclJJ8Rp6UzrL_w   
..          ...                   ...                       ...   
30  A7GIRi5sRlw  2023-06-14T11:53:09Z  UC-lAZVirmOd4TQ5U4lCT4Yg   
31  WbSSjAo9Y5A  2023-06-13T16:55:36Z  UC0dVruAP10W4T3ixx5doiYA   
32  a-1I_MeSwVI  2023-06-14T16:26:57Z  UCyxEbZ6HiT11cDWvCs0vVLQ   
33  IYKNDDLLdk0  2023-06-14T11:37:11Z  UCBdOKRRj1GUXXcISvTipvwQ   
34  KzjUpw20wR8  2023-06-13T16:50:22Z  UCXZPX5qAh1XvaePP8p8_seg   

                                        snippet.title  \
0                    Топ-3 нейросети #айти #нейросети   
1                Лютая сладость от нейросети) Часть 1   
2   Как создать игру если Искусственный Интеллект ...   
3   Нейросеть побила мировой рекорд! #игры #компью...   
4      Нейросеть превращает изображения в АНИМЕ Vol.2   
..                                                ...   
30          Кукольный театр. Король и Шут. Нейросеть.   
31  Нейросеть это будущее которое уже наступило#sh...   
32                         Александр Дюма и нейросеть   
33  НИДЕРЛАНДЫ - ХОРВАТИЯ  #прогноз #лигаконференц...   
34  Manas epic neural networks,  Манас эпос нейрос...   

                                  snippet.description  \
0                                                       
1                                                       
2   Этот туториал поможет вам ответить на вопрос, ...   
3                                                       
4                                                       
..                                                ...   
30  В кукольный театр всех карлик приглашал "Кто в...   
31                                                      
32  В бесплатных уроках рассказываю, как зарабатыв...   
33                                                      
34  kyrgyz #kyrgyzcha_top #ethnic #neuralnetworks ...   

                    snippet.thumbnails.default.url  \
0   https://i.ytimg.com/vi/3OxtnK1m5Sw/default.jpg   
1   https://i.ytimg.com/vi/h6QZtd-Efko/default.jpg   
2   https://i.ytimg.com/vi/HaLZtHN83JE/default.jpg   
3   https://i.ytimg.com/vi/gkDrVYInSaQ/default.jpg   
4   https://i.ytimg.com/vi/P8GXJ7AWqIA/default.jpg   
..                                             ...   
30  https://i.ytimg.com/vi/A7GIRi5sRlw/default.jpg   
31  https://i.ytimg.com/vi/WbSSjAo9Y5A/default.jpg   
32  https://i.ytimg.com/vi/a-1I_MeSwVI/default.jpg   
33  https://i.ytimg.com/vi/IYKNDDLLdk0/default.jpg   
34  https://i.ytimg.com/vi/KzjUpw20wR8/default.jpg   

    snippet.thumbnails.default.width  snippet.thumbnails.default.height  \
0                                120                                 90   
1                                120                                 90   
2                                120                                 90   
3                              

3583 видео (на момент запуска скрипта) получилось по запросу "нейросеть".

In [24]:
# № 7  удаляем дубликаты
df_ne = df_ne.drop_duplicates("id.videoId")
display(df_ne)

kind                         etag        id.kind  \
0   youtube#searchResult  dxv4h9ahuptyF44fL7y24uIK8n0  youtube#video   
1   youtube#searchResult  MwUFhIFJ2sm_-IOduhHKJVw1A2s  youtube#video   
2   youtube#searchResult  C2Fk0U47OES4qJIEUmzvs9pVRfc  youtube#video   
3   youtube#searchResult  BQcCdIuFNFleZ_sQGot_Zq1YNgQ  youtube#video   
4   youtube#searchResult  KR12pERv9PDvaiayuGB_tApP4ZU  youtube#video   
..                   ...                          ...            ...   
21  youtube#searchResult  CGRRehyLe1F7KMgq2X3yboyYRAw  youtube#video   
32  youtube#searchResult  IU86tFFiMqI9ZhWn3pdPo5oV7Eo  youtube#video   
36  youtube#searchResult  wR3X5Oqu18cJkm6HxcD-nH2juvM  youtube#video   
5   youtube#searchResult  t-RxAsjuoY6YCpDc9tjxCdOStVQ  youtube#video   
31  youtube#searchResult  MDVtMD7Y6OrDXoaR4Wnm8z-xGLY  youtube#video   

     id.videoId   snippet.publishedAt         snippet.channelId  \
0   3OxtnK1m5Sw  2022-10-20T22:44:44Z  UCm7bpbe7_H_47ef4dJH0SNw   
1   h6QZtd-Efko  2023-05-04T09:59:13Z  UCf2ZCvUQRwdkFUfebyy0Xag   
2   HaLZtHN83JE  2023-01-26T13:00:15Z  UCRd8_F4-SynaG4sE5lnXv_g   
3   gkDrVYInSaQ  2023-02-10T14:00:24Z  UCUEQBzSZx65-FcVwce8HYNQ   
4   P8GXJ7AWqIA  2022-12-05T07:04:35Z  UCkF05U4kMclJJ8Rp6UzrL_w   
..          ...                   ...                       ...   
21  ZxIN6vCIMe8  2022-06-19T12:36:11Z  UCxMBGX-c9pGG-YOGzdMV1pQ   
32  bhsPu9kaMeA  2023-06-04T09:00:20Z  UCSBeL28cCqIyHFxmCTK1Ejw   
36  WGntxEFMdXo  2023-05-12T12:28:57Z  UCOFQ45F6JIVM1Q7KI-gYT3A   
5   j3od_Wn58DQ  2023-04-25T21:10:31Z  UC-zQQEt2z8kO-lyhaJIJmaQ   
31  WbSSjAo9Y5A  2023-06-13T16:55:36Z  UC0dVruAP10W4T3ixx5doiYA   

                                        snippet.title  \
0                    Топ-3 нейросети #айти #нейросети   
1                Лютая сладость от нейросети) Часть 1   
2   Как создать игру если Искусственный Интеллект ...   
3   Нейросеть побила мировой рекорд! #игры #компью...   
4      Нейросеть превращает изображения в АНИМЕ Vol.2   
..                                                ...   
21                  Нейросеть пояснила за Prey (2006)   
32             Обучаем Нейросеть без Программирования   
36  Господи, как я смеялась 😂 почему меня так не л...   
5   Новая бесплатная нейросеть Kandinsky 2.1 от Сб...   
31  Нейросеть это будущее которое уже наступило#sh...   

                                  snippet.description  \
0                                                       
1                                                       
2   Этот туториал поможет вам ответить на вопрос, ...   
3                                                       
4                                                       
..                                                ...   
21  В сегодняшней телепрограмме мы узнаем, что же ...   
32  Обучаем искусственный интеллект без программир...   
36                                                      
5   Кандинский — обновленная нейросеть, которая ге...   
31                                                      

                    snippet.thumbnails.default.url  \
0   https://i.ytimg.com/vi/3OxtnK1m5Sw/default.jpg   
1   https://i.ytimg.com/vi/h6QZtd-Efko/default.jpg   
2   https://i.ytimg.com/vi/HaLZtHN83JE/default.jpg   
3   https://i.ytimg.com/vi/gkDrVYInSaQ/default.jpg   
4   https://i.ytimg.com/vi/P8GXJ7AWqIA/default.jpg   
..                                             ...   
21  https://i.ytimg.com/vi/ZxIN6vCIMe8/default.jpg   
32  https://i.ytimg.com/vi/bhsPu9kaMeA/default.jpg   
36  https://i.ytimg.com/vi/WGntxEFMdXo/default.jpg   
5   https://i.ytimg.com/vi/j3od_Wn58DQ/default.jpg   
31  https://i.ytimg.com/vi/WbSSjAo9Y5A/default.jpg   

    snippet.thumbnails.default.width  snippet.thumbnails.default.height  \
0                                120                                 90   
1                                120                                 90   
2                                120                                 90   
3                              

После чистки у нас осталось 935 видео, соответствующих нашему запросу.

In [ ]:
# № 8 достали из таблицы с видео их id
videoId_list = df_ne['id.videoId'].to_list()
videoId_list

In [ ]:
# № 9  делаем все тоже самое, что и раньше только с подачей id
request = youtube.videos().list(
    part = ["snippet"              # тут хранятся: дата публикации видео, id канала, название видео, описание видео, tags, categoryId, название канала
            , "contentDetails"     # тут хранится длительность
            , "statistics"         # тут хранятся: количество просмотров, лайков, комментариев
            , "topicDetails"]      # тут хранятся: категории видео
    , id = videoId_list[0:50]
    , maxResults = 50)

response = request.execute()
video_ne = pandas.json_normalize(response['items'])
video_ne

In [ ]:
# № 10
# цикл для прохода по всем id видео из списка videoId_list
for i in range(50, len(videoId_list)+1, 50):
    request = youtube.videos().list(
        part = ["snippet"
                , "contentDetails"
                , "statistics"
                , "topicDetails"]
        , id = videoId_list[i:i+50]
        , maxResults = 50)
    response = request.execute()

    print(f"Итерация №{i}")

    df_additional = pandas.json_normalize(response["items"]) # датафрейм, в котором будут сохранены собранные характеристики каналов с каждой страницы
    video_ne = pandas.concat([video_ne, df_additional])

video_ne

In [28]:
# № 11 удаляем дубликаты из столбца snippet.categoryId
unique_categories = video_ne["snippet.categoryId"].drop_duplicates().tolist()
print(unique_categories)
print(f"Число уникальных категорий в базе: {len(unique_categories)}")

['22', '24', '27', '23', '25', '28', '26', '10', '1', '20', '29', '2', '17', '19', '15']
Число уникальных категорий в базе: 15


In [29]:
# № 12 подаем запрос
request = youtube.videoCategories().list(
    part = "snippet"
    , id = unique_categories)

response = request.execute()

In [30]:
# № 13 работа с id категориями
category_names_df = pandas.json_normalize(response["items"])
category_names_df

kind                         etag  id  \
0   youtube#videoCategory  grPOPYEUUZN3ltuDUGEWlrTR90U   1   
1   youtube#videoCategory  Q0xgUf8BFM8rW3W0R9wNq809xyA   2   
2   youtube#videoCategory  qnpwjh5QlWM5hrnZCvHisquztC4  10   
3   youtube#videoCategory  HyFIixS5BZaoBdkQdLzPdoXWipg  15   
4   youtube#videoCategory  PNU8SwXhjsF90fmkilVohofOi4I  17   
5   youtube#videoCategory  ANnLQyzEA_9m3bMyJXMhKTCOiyg  19   
6   youtube#videoCategory  0Hh6gbZ9zWjnV3sfdZjKB5LQr6E  20   
7   youtube#videoCategory  cHDaaqPDZsJT1FPr1-MwtyIhR28  22   
8   youtube#videoCategory  3Uz364xBbKY50a2s0XQlv-gXJds  23   
9   youtube#videoCategory  0srcLUqQzO7-NGLF7QnhdVzJQmY  24   
10  youtube#videoCategory  bQlQMjmYX7DyFkX4w3kT0osJyIc  25   
11  youtube#videoCategory  Y06N41HP_WlZmeREZvkGF0HW5pg  26   
12  youtube#videoCategory  yBaNkLx4sX9NcDmFgAmxQcV4Y30  27   
13  youtube#videoCategory  Mxy3A-SkmnR7MhJDZRS4DuAIbQA  28   
14  youtube#videoCategory  p3lEirEJApyEkuWpaGEHoF-m-aA  29   

            snippet.title  snippet.assignable         snippet.channelId  
0        Film & Animation                True  UCBR8-60-B28hp2BmDPdntcQ  
1        Autos & Vehicles                True  UCBR8-60-B28hp2BmDPdntcQ  
2                   Music                True  UCBR8-60-B28hp2BmDPdntcQ  
3          Pets & Animals                True  UCBR8-60-B28hp2BmDPdntcQ  
4                  Sports                True  UCBR8-60-B28hp2BmDPdntcQ  
5         Travel & Events                True  UCBR8-60-B28hp2BmDPdntcQ  
6                  Gaming                True  UCBR8-60-B28hp2BmDPdntcQ  
7          People & Blogs                True  UCBR8-60-B28hp2BmDPdntcQ  
8                  Comedy                True  UCBR8-60-B28hp2BmDPdntcQ  
9           Entertainment                True  UCBR8-60-B28hp2BmDPdntcQ  
10        News & Politics                True  UCBR8-60-B28hp2BmDPdntcQ  
11          Howto & Style                True  UCBR8-60-B28hp2BmDPdntcQ  
12              Education                True  UCBR8-60-B28hp2BmDPdntcQ  
13   Science & Technology                True  UCBR8-60-B28hp2BmDPdntcQ  
14  Nonprofits & Activism                True  UCBR8-60-B28hp2BmDPdntcQ

In [31]:
category_names_df["id"]  # dtype: object меняем во всем столбце на числа
category_names_df["id"] = category_names_df["id"].astype("int")

In [32]:
category_names_df["id"]

0      1
1      2
2     10
3     15
4     17
5     19
6     20
7     22
8     23
9     24
10    25
11    26
12    27
13    28
14    29
Name: id, dtype: int64

In [33]:
# №14 замена сквозной нумерации в таблице с id на номера самих id
category_names_df.index = category_names_df["id"].to_list()
category_names_df

kind                         etag  id  \
1   youtube#videoCategory  grPOPYEUUZN3ltuDUGEWlrTR90U   1   
2   youtube#videoCategory  Q0xgUf8BFM8rW3W0R9wNq809xyA   2   
10  youtube#videoCategory  qnpwjh5QlWM5hrnZCvHisquztC4  10   
15  youtube#videoCategory  HyFIixS5BZaoBdkQdLzPdoXWipg  15   
17  youtube#videoCategory  PNU8SwXhjsF90fmkilVohofOi4I  17   
19  youtube#videoCategory  ANnLQyzEA_9m3bMyJXMhKTCOiyg  19   
20  youtube#videoCategory  0Hh6gbZ9zWjnV3sfdZjKB5LQr6E  20   
22  youtube#videoCategory  cHDaaqPDZsJT1FPr1-MwtyIhR28  22   
23  youtube#videoCategory  3Uz364xBbKY50a2s0XQlv-gXJds  23   
24  youtube#videoCategory  0srcLUqQzO7-NGLF7QnhdVzJQmY  24   
25  youtube#videoCategory  bQlQMjmYX7DyFkX4w3kT0osJyIc  25   
26  youtube#videoCategory  Y06N41HP_WlZmeREZvkGF0HW5pg  26   
27  youtube#videoCategory  yBaNkLx4sX9NcDmFgAmxQcV4Y30  27   
28  youtube#videoCategory  Mxy3A-SkmnR7MhJDZRS4DuAIbQA  28   
29  youtube#videoCategory  p3lEirEJApyEkuWpaGEHoF-m-aA  29   

            snippet.title  snippet.assignable         snippet.channelId  
1        Film & Animation                True  UCBR8-60-B28hp2BmDPdntcQ  
2        Autos & Vehicles                True  UCBR8-60-B28hp2BmDPdntcQ  
10                  Music                True  UCBR8-60-B28hp2BmDPdntcQ  
15         Pets & Animals                True  UCBR8-60-B28hp2BmDPdntcQ  
17                 Sports                True  UCBR8-60-B28hp2BmDPdntcQ  
19        Travel & Events                True  UCBR8-60-B28hp2BmDPdntcQ  
20                 Gaming                True  UCBR8-60-B28hp2BmDPdntcQ  
22         People & Blogs                True  UCBR8-60-B28hp2BmDPdntcQ  
23                 Comedy                True  UCBR8-60-B28hp2BmDPdntcQ  
24          Entertainment                True  UCBR8-60-B28hp2BmDPdntcQ  
25        News & Politics                True  UCBR8-60-B28hp2BmDPdntcQ  
26          Howto & Style                True  UCBR8-60-B28hp2BmDPdntcQ  
27              Education                True  UCBR8-60-B28hp2BmDPdntcQ  
28   Science & Technology                True  UCBR8-60-B28hp2BmDPdntcQ  
29  Nonprofits & Activism                True  UCBR8-60-B28hp2BmDPdntcQ

In [34]:
# № 15 столбец новый куда поместим категории видео
video_ne["categoryName"] = 0
video_ne

kind                         etag           id  \
0   youtube#video  3Quzg6U7m27RGULnMGXX84MXtcU  3OxtnK1m5Sw   
1   youtube#video  LYu0uYxoQflIJrgN2XTBoV8wsfg  h6QZtd-Efko   
2   youtube#video  ZnldGYX5-r7vEVNcipKSS1THs0Y  HaLZtHN83JE   
3   youtube#video  CvlkebysmY6o4u1LMKjKOME_h3o  gkDrVYInSaQ   
4   youtube#video  XX4xhLYYZg4PDAYQzzg21BP0390  P8GXJ7AWqIA   
..            ...                          ...          ...   
33  youtube#video  TzsyAHjJDdj6pRmeA62DTobCxHY  ZxIN6vCIMe8   
34  youtube#video  LTzwbokMFakiAFMnqOUnkK8fAWY  bhsPu9kaMeA   
35  youtube#video  1kfjmZMV_ZElSzuoyfzmLYvCwcc  WGntxEFMdXo   
36  youtube#video  7sN1DWYCRMUEA4p4AKPamghqgNU  j3od_Wn58DQ   
37  youtube#video  f0lt1hOYMrLB0NlfVA0s-4tpLsM  WbSSjAo9Y5A   

     snippet.publishedAt         snippet.channelId  \
0   2022-10-20T22:44:44Z  UCm7bpbe7_H_47ef4dJH0SNw   
1   2023-05-04T09:59:13Z  UCf2ZCvUQRwdkFUfebyy0Xag   
2   2023-01-26T13:00:15Z  UCRd8_F4-SynaG4sE5lnXv_g   
3   2023-02-10T14:00:24Z  UCUEQBzSZx65-FcVwce8HYNQ   
4   2022-12-05T07:04:35Z  UCkF05U4kMclJJ8Rp6UzrL_w   
..                   ...                       ...   
33  2022-06-19T12:36:11Z  UCxMBGX-c9pGG-YOGzdMV1pQ   
34  2023-06-04T09:00:20Z  UCSBeL28cCqIyHFxmCTK1Ejw   
35  2023-05-12T12:28:57Z  UCOFQ45F6JIVM1Q7KI-gYT3A   
36  2023-04-25T21:10:31Z  UC-zQQEt2z8kO-lyhaJIJmaQ   
37  2023-06-13T16:55:36Z  UC0dVruAP10W4T3ixx5doiYA   

                                        snippet.title  \
0                    Топ-3 нейросети #айти #нейросети   
1                Лютая сладость от нейросети) Часть 1   
2   Как создать игру если Искусственный Интеллект ...   
3   Нейросеть побила мировой рекорд! #игры #компью...   
4      Нейросеть превращает изображения в АНИМЕ Vol.2   
..                                                ...   
33                  Нейросеть пояснила за Prey (2006)   
34             Обучаем Нейросеть без Программирования   
35  Господи, как я смеялась 😂 почему меня так не л...   
36  Новая бесплатная нейросеть Kandinsky 2.1 от Сб...   
37  Нейросеть это будущее которое уже наступило#sh...   

                                  snippet.description  \
0                                                       
1                                                       
2   Этот туториал поможет вам ответить на вопрос, ...   
3                                                       
4                                                       
..                                                ...   
33  В сегодняшней телепрограмме мы узнаем, что же ...   
34  Обучаем искусственный интеллект без программир...   
35                                                      
36  Кандинский — обновленная нейросеть, которая ге...   
37                                                      

                    snippet.thumbnails.default.url  \
0   https://i.ytimg.com/vi/3OxtnK1m5Sw/default.jpg   
1   https://i.ytimg.com/vi/h6QZtd-Efko/default.jpg   
2   https://i.ytimg.com/vi/HaLZtHN83JE/default.jpg   
3   https://i.ytimg.com/vi/gkDrVYInSaQ/default.jpg   
4   https://i.ytimg.com/vi/P8GXJ7AWqIA/default.jpg   
..                                             ...   
33  https://i.ytimg.com/vi/ZxIN6vCIMe8/default.jpg   
34  https://i.ytimg.com/vi/bhsPu9kaMeA/default.jpg   
35  https://i.ytimg.com/vi/WGntxEFMdXo/default.jpg   
36  https://i.ytimg.com/vi/j3od_Wn58DQ/default.jpg   
37  https://i.ytimg.com/vi/WbSSjAo9Y5A/default.jpg   

    snippet.thumbnails.default.width  snippet.thumbnails.default.height  ...  \
0                                120                                 90  ...   
1                                120                                 90  ...   
2                                120                                 90  ...   
3                                120                                 90  ...   
4                                120                                 90  ...   
..                               ...                                ...  ...   
33                          

In [35]:
video_ne["snippet.categoryId"] = video_ne["snippet.categoryId"].astype("int")
video_ne.loc[video_ne["snippet.categoryId"] == 1, "categoryName"] = category_names_df["snippet.title"][1]
video_ne[video_ne["snippet.categoryId"] == 1]

kind                         etag           id  \
6   youtube#video  dvQiuo1jEA1If8qvI7kzYuhGvqE  hd8aREaDxmM   
7   youtube#video  0hS1Zim8NWBgVFYj5cKAlvWcsrY  yYUw700iVpY   
15  youtube#video  -9TkrqqxQVzQXz0gUHGcefE-CMA  p3DyrMVIUjQ   
24  youtube#video  Tgy0kHg1KFEQ6UViqhb44Im2Yzw  Qab3uqKlsh0   
3   youtube#video  8CulfCbX5BaXUuPorJF0gTfhGDc  z7Bxm_Cx_SU   
37  youtube#video  OvBL_Z3Y0joRfKdXRgH2lHmbizU  33ucNdBxgyw   
4   youtube#video  LSwiEmqn9FjKCrsKHq6-TYE4SJ0  I_HKDGRCzL0   
9   youtube#video  V8xZUbTe4DJYlkPJkLo6pLOQ9EY  _TG5i1jkwMM   
1   youtube#video  rk1m8NgzCtgoEUxPaVclCZHWcyM  vvOfg3xoTMs   
14  youtube#video  MDGjGRmTTSjWJQuWav9-hKNZCbo  gryvrexmcfM   
28  youtube#video  NpcEixLniLyZMFnLj6QG8skwPaE  j_tbsbvVF8E   
0   youtube#video  OXQYZTRNBQ0z-J7q7WhB-BepxVk  iKRmJfeuf94   
40  youtube#video  Gi77jNpEwQkMqCx9su2GM5xSWlw  ocB9bMacXTs   
42  youtube#video  x0ifkoykFZQBUup8QZ9TqDh_aRs  pX4EfTtkiOY   
43  youtube#video  Vk7VQDl5cP5ggnJBpHgGWw3xVuk  P4Ark2KAdgQ   
24  youtube#video  wZYarCZlEfkhUA0CNcMkGsnA7uk  mY47r454rmU   
18  youtube#video  7Iv2GxLMvL4GOjLqtWS_wqg0J7I  malTSu8gAnI   
29  youtube#video  mumlmSETF9cmWO9W6490yq2Ne0U  k8QZ2AmjyeY   
5   youtube#video  7aR4sCuj10C-ICaG7QnDcFPWmDY  oiX9Vr2rfGM   
14  youtube#video  Za99m6vs1-48OxdjSMOYnXfG1WI  Kbu-7AL3hEA   
26  youtube#video  9WAtMGf6GP5uOwTw2yb8OfIsz3E  h3kNCZ0KLuM   
34  youtube#video  -RceBYmRxMwLLVI9QYiBk32AWJo  ZgZhAXo6yEA   
41  youtube#video  -sWeFDPldRdnkoUOjJG389QEwpU  eKVTWVH_oVo   
8   youtube#video  dW6OA2v63ZDDM2Idv54vJOius4M  FW5_EvDoX4k   
16  youtube#video  sGj1k9wJtNjJJaUKLWV1kQD_u3A  djsTCdhb4tw   
29  youtube#video  CyEwlVKtNTrFr6sJ6TP-4a3LK1g  ndWWV-LNVXo   
34  youtube#video  S1dAfB4PJ_Dl2bWpoTKaUjFlRms  cpZWlf_rJh8   
40  youtube#video  3KnilJQGdVWGb84gF0Gut0qXPcg  xjyBKGodh6w   
31  youtube#video  FRq715O3zfPdUcHr91oG-JTRoPE  Y7zJnZpnO9Y   
32  youtube#video  SE8yQtIBylGaCtr4By-j21P4TzU  PQNd2vTM_0Q   
41  youtube#video  QjM4N5oWCZjdhpYE84KOJ84GQlE  loNDT_YHRcQ   
6   youtube#video  02XDTAgVHlZqQUhgu-Qv7qO86AQ  Eptezfsl81U   
7   youtube#video  r5H3tD3aFxmB46JFI0HPt9Oy90I  5iEngc8kH84   
24  youtube#video  Mw6I3NFhY-hodFHvTLdA_zcLDxk  Pbve8CVc_3I   
33  youtube#video  TzsyAHjJDdj6pRmeA62DTobCxHY  ZxIN6vCIMe8   

     snippet.publishedAt         snippet.channelId  \
6   2022-12-12T14:00:31Z  UC6nyZJt0-94zP2IyZ46-9zg   
7   2023-03-21T15:33:54Z  UCUmlB9SwrBVevETZV0wrVRw   
15  2023-06-13T13:22:06Z  UC9NTnR3ZqiN9QjCu0UPWCHA   
24  2023-06-12T17:45:35Z  UCDiDMw8TI6wEUqiPGm74Qmg   
3   2022-11-12T09:00:38Z  UCZH2Lzc8KxixJOgXF50g_VQ   
37  2023-03-21T13:00:03Z  UCEeieeMrsmM8TllpJpYvKJA   
4   2023-06-14T11:48:56Z  UCEsmdt936GoKO6zAUkq7cGw   
9   2023-06-11T05:00:15Z  UCfxlHMwLCdvZ8MzJu9pAlBQ   
1   2023-06-06T13:59:54Z  UCGNIf8qTK5lgwp0yysGKc7g   
14  2023-05-10T17:00:21Z  UCEG-C_q98ya6fKMeSYJAMbg   
28  2023-04-12T15:00:40Z  UCdIE1S-7ZiMJMbWJ3rlb5Qw   
0   2023-03-29T12:08:00Z  UCUmlB9SwrBVevETZV0wrVRw   
40  2023-03-13T11:30:15Z  UCSV96ZPg2c4lg6Gu5sJl77g   
42  2023-02-26T10:00:35Z  UCeAvLWwxxbustQCBLU_s81A   
43  2023-02-27T18:06:21Z  UCWE9a9b6SJasgWK2jJxVUww   
24  2022-12-20T11:48:47Z  UCc0AOE8mAXSKwzPJ1kg2yCQ   
18  2022-11-06T12:27:25Z  UCj8JGFwl8SmcBBW9C6LxhHA   
29  2023-03-10T21:18:57Z  UCeAvLWwxxbustQCBLU_s81A   
5   2022-08-16T09:35:59Z  UCyND9OVKZ8e5QDdmundREgQ   
14  2023-03-31T12:00:20Z  UCUmlB9SwrBVevETZV0wrVRw   
26  2022-11-09T19:45:18Z  UC54oHxLSrDdumdV_eeHmIQg   
34  2023-05-30T12:00:09Z  UC1PVohA0bU5uRILb0iiljIg   
41  2022-11-22T12:00:28Z  UCUmlB9SwrBVevETZV0wrVRw   
8   2023-03-04T12:00:09Z  UCUmlB9SwrBVevETZV0wrVRw   
16  2023-03-13T17:26:39Z  UCSV96ZPg2c4lg6Gu5sJl77g   
29  2017-10-31T11:32:28Z  UCxtTXA5DU1bZHLLomu83zVw   
34  2023-05-12T12:00:27Z  UC3L93d3zybiUGE0wwbfC2iA   
40  2023-05-15T16:14:00Z  UCdIE1S-7ZiMJMbWJ3rlb5Qw   
31  2021-08-03T09:24:37Z  UCc0AOE8mAXSKwzPJ1kg2yCQ   
32  2023-03-17T11:05:15Z  UCWkZTUaVTDz0obxxagQQkQg   
41  2023-03-17T13:07:07Z  UCzxhqZLzCGJRuRQiAwJtR-g   
6   2023-01-26T1

In [36]:
# цикл для того, чтобы перезаписать расшифровки всех категорий id
for i in category_names_df.index:
    video_ne.loc[video_ne["snippet.categoryId"] == i, "categoryName"] = category_names_df["snippet.title"][i]
video_ne

kind                         etag           id  \
0   youtube#video  3Quzg6U7m27RGULnMGXX84MXtcU  3OxtnK1m5Sw   
1   youtube#video  LYu0uYxoQflIJrgN2XTBoV8wsfg  h6QZtd-Efko   
2   youtube#video  ZnldGYX5-r7vEVNcipKSS1THs0Y  HaLZtHN83JE   
3   youtube#video  CvlkebysmY6o4u1LMKjKOME_h3o  gkDrVYInSaQ   
4   youtube#video  XX4xhLYYZg4PDAYQzzg21BP0390  P8GXJ7AWqIA   
..            ...                          ...          ...   
33  youtube#video  TzsyAHjJDdj6pRmeA62DTobCxHY  ZxIN6vCIMe8   
34  youtube#video  LTzwbokMFakiAFMnqOUnkK8fAWY  bhsPu9kaMeA   
35  youtube#video  1kfjmZMV_ZElSzuoyfzmLYvCwcc  WGntxEFMdXo   
36  youtube#video  7sN1DWYCRMUEA4p4AKPamghqgNU  j3od_Wn58DQ   
37  youtube#video  f0lt1hOYMrLB0NlfVA0s-4tpLsM  WbSSjAo9Y5A   

     snippet.publishedAt         snippet.channelId  \
0   2022-10-20T22:44:44Z  UCm7bpbe7_H_47ef4dJH0SNw   
1   2023-05-04T09:59:13Z  UCf2ZCvUQRwdkFUfebyy0Xag   
2   2023-01-26T13:00:15Z  UCRd8_F4-SynaG4sE5lnXv_g   
3   2023-02-10T14:00:24Z  UCUEQBzSZx65-FcVwce8HYNQ   
4   2022-12-05T07:04:35Z  UCkF05U4kMclJJ8Rp6UzrL_w   
..                   ...                       ...   
33  2022-06-19T12:36:11Z  UCxMBGX-c9pGG-YOGzdMV1pQ   
34  2023-06-04T09:00:20Z  UCSBeL28cCqIyHFxmCTK1Ejw   
35  2023-05-12T12:28:57Z  UCOFQ45F6JIVM1Q7KI-gYT3A   
36  2023-04-25T21:10:31Z  UC-zQQEt2z8kO-lyhaJIJmaQ   
37  2023-06-13T16:55:36Z  UC0dVruAP10W4T3ixx5doiYA   

                                        snippet.title  \
0                    Топ-3 нейросети #айти #нейросети   
1                Лютая сладость от нейросети) Часть 1   
2   Как создать игру если Искусственный Интеллект ...   
3   Нейросеть побила мировой рекорд! #игры #компью...   
4      Нейросеть превращает изображения в АНИМЕ Vol.2   
..                                                ...   
33                  Нейросеть пояснила за Prey (2006)   
34             Обучаем Нейросеть без Программирования   
35  Господи, как я смеялась 😂 почему меня так не л...   
36  Новая бесплатная нейросеть Kandinsky 2.1 от Сб...   
37  Нейросеть это будущее которое уже наступило#sh...   

                                  snippet.description  \
0                                                       
1                                                       
2   Этот туториал поможет вам ответить на вопрос, ...   
3                                                       
4                                                       
..                                                ...   
33  В сегодняшней телепрограмме мы узнаем, что же ...   
34  Обучаем искусственный интеллект без программир...   
35                                                      
36  Кандинский — обновленная нейросеть, которая ге...   
37                                                      

                    snippet.thumbnails.default.url  \
0   https://i.ytimg.com/vi/3OxtnK1m5Sw/default.jpg   
1   https://i.ytimg.com/vi/h6QZtd-Efko/default.jpg   
2   https://i.ytimg.com/vi/HaLZtHN83JE/default.jpg   
3   https://i.ytimg.com/vi/gkDrVYInSaQ/default.jpg   
4   https://i.ytimg.com/vi/P8GXJ7AWqIA/default.jpg   
..                                             ...   
33  https://i.ytimg.com/vi/ZxIN6vCIMe8/default.jpg   
34  https://i.ytimg.com/vi/bhsPu9kaMeA/default.jpg   
35  https://i.ytimg.com/vi/WGntxEFMdXo/default.jpg   
36  https://i.ytimg.com/vi/j3od_Wn58DQ/default.jpg   
37  https://i.ytimg.com/vi/WbSSjAo9Y5A/default.jpg   

    snippet.thumbnails.default.width  snippet.thumbnails.default.height  ...  \
0                                120                                 90  ...   
1                                120                                 90  ...   
2                                120                                 90  ...   
3                                120                                 90  ...   
4                                120                                 90  ...   
..                               ...                                ...  ...   
33                          

In [37]:
# № 16  сквозной индекс для итоговой таблицы
video_ne.index = range(1, len(video_ne)+1)
display(video_ne)

kind                         etag           id  \
1    youtube#video  3Quzg6U7m27RGULnMGXX84MXtcU  3OxtnK1m5Sw   
2    youtube#video  LYu0uYxoQflIJrgN2XTBoV8wsfg  h6QZtd-Efko   
3    youtube#video  ZnldGYX5-r7vEVNcipKSS1THs0Y  HaLZtHN83JE   
4    youtube#video  CvlkebysmY6o4u1LMKjKOME_h3o  gkDrVYInSaQ   
5    youtube#video  XX4xhLYYZg4PDAYQzzg21BP0390  P8GXJ7AWqIA   
..             ...                          ...          ...   
884  youtube#video  TzsyAHjJDdj6pRmeA62DTobCxHY  ZxIN6vCIMe8   
885  youtube#video  LTzwbokMFakiAFMnqOUnkK8fAWY  bhsPu9kaMeA   
886  youtube#video  1kfjmZMV_ZElSzuoyfzmLYvCwcc  WGntxEFMdXo   
887  youtube#video  7sN1DWYCRMUEA4p4AKPamghqgNU  j3od_Wn58DQ   
888  youtube#video  f0lt1hOYMrLB0NlfVA0s-4tpLsM  WbSSjAo9Y5A   

      snippet.publishedAt         snippet.channelId  \
1    2022-10-20T22:44:44Z  UCm7bpbe7_H_47ef4dJH0SNw   
2    2023-05-04T09:59:13Z  UCf2ZCvUQRwdkFUfebyy0Xag   
3    2023-01-26T13:00:15Z  UCRd8_F4-SynaG4sE5lnXv_g   
4    2023-02-10T14:00:24Z  UCUEQBzSZx65-FcVwce8HYNQ   
5    2022-12-05T07:04:35Z  UCkF05U4kMclJJ8Rp6UzrL_w   
..                    ...                       ...   
884  2022-06-19T12:36:11Z  UCxMBGX-c9pGG-YOGzdMV1pQ   
885  2023-06-04T09:00:20Z  UCSBeL28cCqIyHFxmCTK1Ejw   
886  2023-05-12T12:28:57Z  UCOFQ45F6JIVM1Q7KI-gYT3A   
887  2023-04-25T21:10:31Z  UC-zQQEt2z8kO-lyhaJIJmaQ   
888  2023-06-13T16:55:36Z  UC0dVruAP10W4T3ixx5doiYA   

                                         snippet.title  \
1                     Топ-3 нейросети #айти #нейросети   
2                 Лютая сладость от нейросети) Часть 1   
3    Как создать игру если Искусственный Интеллект ...   
4    Нейросеть побила мировой рекорд! #игры #компью...   
5       Нейросеть превращает изображения в АНИМЕ Vol.2   
..                                                 ...   
884                  Нейросеть пояснила за Prey (2006)   
885             Обучаем Нейросеть без Программирования   
886  Господи, как я смеялась 😂 почему меня так не л...   
887  Новая бесплатная нейросеть Kandinsky 2.1 от Сб...   
888  Нейросеть это будущее которое уже наступило#sh...   

                                   snippet.description  \
1                                                        
2                                                        
3    Этот туториал поможет вам ответить на вопрос, ...   
4                                                        
5                                                        
..                                                 ...   
884  В сегодняшней телепрограмме мы узнаем, что же ...   
885  Обучаем искусственный интеллект без программир...   
886                                                      
887  Кандинский — обновленная нейросеть, которая ге...   
888                                                      

                     snippet.thumbnails.default.url  \
1    https://i.ytimg.com/vi/3OxtnK1m5Sw/default.jpg   
2    https://i.ytimg.com/vi/h6QZtd-Efko/default.jpg   
3    https://i.ytimg.com/vi/HaLZtHN83JE/default.jpg   
4    https://i.ytimg.com/vi/gkDrVYInSaQ/default.jpg   
5    https://i.ytimg.com/vi/P8GXJ7AWqIA/default.jpg   
..                                              ...   
884  https://i.ytimg.com/vi/ZxIN6vCIMe8/default.jpg   
885  https://i.ytimg.com/vi/bhsPu9kaMeA/default.jpg   
886  https://i.ytimg.com/vi/WGntxEFMdXo/default.jpg   
887  https://i.ytimg.com/vi/j3od_Wn58DQ/default.jpg   
888  https://i.ytimg.com/vi/WbSSjAo9Y5A/default.jpg   

     snippet.thumbnails.default.width  snippet.thumbnails.default.height  ...  \
1                                 120                                 90  ...   
2                                 120                                 90  ...   
3                                 120                                 90  ...   
4                                 120                                 90  ...   
5                                 120                                 90  ...   
..                                ...      

In [38]:
# № 17 теперь наша задача собрать дизлайки для создания переменной вовлеченность, через собственно-ручно созданную функцию  dislikes()
userAgent = ["Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.150 Safari/537.36 RuxitSynthetic/1.0 v3653660536 t3060034875834570496 ath318f403d altpub cvcv=2 smf=0"
             , "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.150 Safari/537.36 RuxitSynthetic/1.0 v7638580655095651333 t3369586550842097326 ath1fb31b7a altpriv cvcv=2 smf=0"
             , "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.150 Safari/537.36 RuxitSynthetic/1.0 v4022270135305624536 t210370085878917362 ath259cea6f altpriv cvcv=2 smf=0"
             , "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.66 Safari/537.36 RuxitSynthetic/1.0 v1605625219501687423 t5534040873462942519 ath5ee645e0 altpriv cvcv=2 smf=0"
             , "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36 RuxitSynthetic/1.0 v9045054735720176711 t8055053424576152628 ath4b3726d5 altpriv cvcv=2 smf=0"
             , "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.150 Safari/537.36 RuxitSynthetic/1.0 v4774994596211760778 t5534040873462942519 ath5ee645e0 altpriv cvcv=2 smf=0"
             , "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36 RuxitSynthetic/1.0 v1418982706 t79976092658796286 athfa3c3975 altpub cvcv=2 smf=0"
             , "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.150 Safari/537.36 RuxitSynthetic/1.0 v4650801409596051110 t3369586550842097326 ath1fb31b7a altpriv cvcv=2 smf=0"]

def dislikes(videoId, likeCount = None):
    url = "https://returnyoutubedislikeapi.com/votes?videoId=" + videoId

    if likeCount != None:
        url.append("&likeCount=" + str(likeCount))

    headers = {'authority': 'returnyoutubedislikeapi.com'
               , 'cache-control': 'max-age=0'
               , 'sec-ch-ua': '" Not A;Brand";v="99", "Chromium";v="98", "Microsoft Edge";v="98"'
               , 'sec-ch-ua-mobile': '?0'
               , 'sec-ch-ua-platform': '"Windows"'
               , 'upgrade-insecure-requests': '1'
               , 'user-agent': random.choice(userAgent)
               , 'accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9'
               , 'sec-fetch-site': 'none'
               , 'sec-fetch-mode': 'navigate'
               , 'sec-fetch-user': '?1'
               , 'sec-fetch-dest': 'document'
               , 'accept-language': 'ru,en;q=0.9,en-GB;q=0.8,en-US;q=0.7'}

    response = requests.request("GET", url, headers = headers)
    return response.json()["dislikes"]

In [ ]:
## 17.2 собираем через цикл дизлайки
indeseS = video_ne.index
# пустой список errorS для ошибок
errorS = []
for j in indeseS:
    time.sleep(1)
    print(f"Строка №{j}")
    try:
        video_ne.loc[j, "statistics.dislikesCount"] = dislikes(video_ne["id"][j]) # новый столбец "statistics.dislikesCount" в датафрейме, где количество дизлайков у каждого видео
    except:
        print(f"Ошибка в строке {j}")
        errorS.append(j)
video_ne

(возможно, что код будет отрабатывать очень долго в этом чанке, минут 8, поэтому не бейте палками, он очень старается)

Теперь наша задачи состоит в том, чтобы вытащить года видео и посчитать вовлеченность для решения задач. Вовлеченность мы определяем как (лайки+дизлайки+комментарии)/количество просмотров.

In [40]:
# № 18 переменная года выпуска
video_ne["year"] = video_ne["snippet.publishedAt"].apply(lambda yyyy: yyyy[:4])
video_ne["year"] = video_ne["year"].astype("int")

In [41]:
# № 19 работаем с  новой переменной, вовлеченностью в контент (видео)
# ищем и заменяем пропуски, зачастую это связано с тем, что комментарии отключены, то есть их нет у видео,
#  поэтому было принято решение перекодировать их в 0, чтобы не терять наблюдения, тоже самое с лайками
video_ne["statistics.commentCount"] = video_ne["statistics.commentCount"].fillna(0)
video_ne["statistics.likeCount"] = video_ne["statistics.likeCount"].fillna(0)

video_ne["statistics.commentCount"] = video_ne["statistics.commentCount"].astype("int")
video_ne["statistics.likeCount"] = video_ne["statistics.likeCount"].astype("int")
video_ne["statistics.dislikesCount"] = video_ne["statistics.dislikesCount"].astype("int")
video_ne["statistics.viewCount"] = video_ne["statistics.viewCount"].astype("int")

video_ne["involvement"] = round((video_ne["statistics.likeCount"] + video_ne["statistics.dislikesCount"] + video_ne["statistics.commentCount"]) / video_ne["statistics.viewCount"], 2)

In [42]:
video_ne[video_ne["involvement"] == math.inf] #проверили нет ли случайно бесконечности в таблице, ура ее нет

Empty DataFrame
Columns: [kind, etag, id, snippet.publishedAt, snippet.channelId, snippet.title, snippet.description, snippet.thumbnails.default.url, snippet.thumbnails.default.width, snippet.thumbnails.default.height, snippet.thumbnails.medium.url, snippet.thumbnails.medium.width, snippet.thumbnails.medium.height, snippet.thumbnails.high.url, snippet.thumbnails.high.width, snippet.thumbnails.high.height, snippet.thumbnails.standard.url, snippet.thumbnails.standard.width, snippet.thumbnails.standard.height, snippet.thumbnails.maxres.url, snippet.thumbnails.maxres.width, snippet.thumbnails.maxres.height, snippet.channelTitle, snippet.categoryId, snippet.liveBroadcastContent, snippet.localized.title, snippet.localized.description, contentDetails.duration, contentDetails.dimension, contentDetails.definition, contentDetails.caption, contentDetails.licensedContent, contentDetails.projection, statistics.viewCount, statistics.likeCount, statistics.favoriteCount, statistics.commentCount, topicDetails.topicCategories, snippet.tags, snippet.defaultLanguage, snippet.defaultAudioLanguage, contentDetails.regionRestriction.blocked, contentDetails.contentRating.ytRating, categoryName, statistics.dislikesCount, year, involvement]
Index: []

[0 rows x 47 columns]

In [43]:
# № 20 наконец выбираем только нужные нам категории для будущей готовой таблицы
category = ["id"
             , "snippet.publishedAt"
             , "snippet.channelId"
             , "snippet.title"
             , "snippet.description"
             , "snippet.tags"
             , "snippet.categoryId"
             , "topicDetails.topicCategories"
             , "snippet.channelTitle"
             , "contentDetails.duration"
             , "statistics.viewCount"
             , "statistics.likeCount"
             , "statistics.commentCount"
             , "categoryName"
             , "statistics.dislikesCount"
             , "involvement"
             , "year"]
data_neiro = pandas.DataFrame()

for i in video_ne:
    if i in category:
        data_neiro [i] = video_ne[i]

data_neiro # готовые собранные данные записанные в новую таблицу для дальнейшей работы

id   snippet.publishedAt         snippet.channelId  \
1    3OxtnK1m5Sw  2022-10-20T22:44:44Z  UCm7bpbe7_H_47ef4dJH0SNw   
2    h6QZtd-Efko  2023-05-04T09:59:13Z  UCf2ZCvUQRwdkFUfebyy0Xag   
3    HaLZtHN83JE  2023-01-26T13:00:15Z  UCRd8_F4-SynaG4sE5lnXv_g   
4    gkDrVYInSaQ  2023-02-10T14:00:24Z  UCUEQBzSZx65-FcVwce8HYNQ   
5    P8GXJ7AWqIA  2022-12-05T07:04:35Z  UCkF05U4kMclJJ8Rp6UzrL_w   
..           ...                   ...                       ...   
884  ZxIN6vCIMe8  2022-06-19T12:36:11Z  UCxMBGX-c9pGG-YOGzdMV1pQ   
885  bhsPu9kaMeA  2023-06-04T09:00:20Z  UCSBeL28cCqIyHFxmCTK1Ejw   
886  WGntxEFMdXo  2023-05-12T12:28:57Z  UCOFQ45F6JIVM1Q7KI-gYT3A   
887  j3od_Wn58DQ  2023-04-25T21:10:31Z  UC-zQQEt2z8kO-lyhaJIJmaQ   
888  WbSSjAo9Y5A  2023-06-13T16:55:36Z  UC0dVruAP10W4T3ixx5doiYA   

                                         snippet.title  \
1                     Топ-3 нейросети #айти #нейросети   
2                 Лютая сладость от нейросети) Часть 1   
3    Как создать игру если Искусственный Интеллект ...   
4    Нейросеть побила мировой рекорд! #игры #компью...   
5       Нейросеть превращает изображения в АНИМЕ Vol.2   
..                                                 ...   
884                  Нейросеть пояснила за Prey (2006)   
885             Обучаем Нейросеть без Программирования   
886  Господи, как я смеялась 😂 почему меня так не л...   
887  Новая бесплатная нейросеть Kandinsky 2.1 от Сб...   
888  Нейросеть это будущее которое уже наступило#sh...   

                                   snippet.description  \
1                                                        
2                                                        
3    Этот туториал поможет вам ответить на вопрос, ...   
4                                                        
5                                                        
..                                                 ...   
884  В сегодняшней телепрограмме мы узнаем, что же ...   
885  Обучаем искусственный интеллект без программир...   
886                                                      
887  Кандинский — обновленная нейросеть, которая ге...   
888                                                      

                      snippet.channelTitle  snippet.categoryId  \
1                              YCLA-Coding                  22   
2                            Добрый Джонни                  22   
3                             GameDev Club                  24   
4                  Роман Сакутин - GameDev                  27   
5                                amindblow                  22   
..                                     ...                 ...   
884                            Qwetux 2009                   1   
885  Айтигенио: международная онлайн-школа                  20   
886                            Анна Булгак                  10   
887               Нейросетевая Лаборатория                  28   
888                           БумНейросети                  22   

    contentDetails.duration  statistics.viewCount  statistics.likeCount  \
1                     PT36S              11005576                801306   
2                     PT59S              10420652                717458   
3                     PT54S               9157840                420422   
4                     PT42S               6481887                320103   
5                     PT21S               6269361                293735   
..                      ...                   ...                   ...   
884                 PT4M33S                  3544                   328   
885                 PT6M23S                  3918                    39   
886                   PT31S                  2667                    63   
887                 PT8M24S                  2139                   104   
888                   PT33S                   198                     5   

     statistics.commentCount  \
1                       2094   
2                       1137   
3            

# Ура, пристпупаем к анализу данных!
Мы начали со второй задачи, так как когда-то она показалась нам проще.
#### Задача 2. Выяснить связана ли вовлеченность и год выпуска видео

Для решения этой задачи мы решили прибегнуть к коэфициэнту корреляции Пирсона для нахождения парных связей.

 **Гипотезы:**

 H0: вовлеченность в контент (видео) не связана с годом выпуска видео. Коэффициент corr = 0

 H1: вовлеченность в контент (видео) связана с годом выпуска видео (чем позже опубликовано видео (2023 год), тем выше уровень вовлеченности так, как популярность нейросеть стала проявлять в массовой культуре относительно недавно). Коэффициент corr не равен 0.

In [44]:
# № 22 строим графики, чтобы проанализировать переменные визуально
fig221 = px.histogram(data_neiro
                      , x = "involvement"
                      , template = "simple_white"
                      , title = "Распределение видео по запросу 'нейросеть' ")
fig221.update_layout(yaxis_title = "Частота"
                     , xaxis_title = "Вовлеченность в контент, баллы"
                     , title = {"font": dict(size = 15)
                                , "x": 0.5})
fig221.update_traces(hovertemplate = "Частота: %{y}")

fig221.show()

Можно заметить, что график похож на нормальное смещенное распределение,за исключением выбросов.
Скорее всего в дальнейшем при подсчете корреляции, мы избавимся от них.
Посмотрим внимательнее на выбросы до этого.

In [45]:
fig222 = px.box(data_neiro
                , y = "involvement"
                , points = "all"
                , template = "simple_white"
                , title = "Ящик с усами для переменной 'Вовлеченность в контент (видео)'")
fig222.update_layout(yaxis_title = "Вовлеченность в контент (видео), баллы"
                     , title = {"font": dict(size = 15)
                                , "x": 0.5})

fig222.show()

In [46]:
fig231 = px.histogram(data_neiro
                      , x = "year"
                      , template = "simple_white"
                      , title = "Частотное распределение видео по запросу 'нейросеть' <br>по годам выпуска")
fig231.update_layout(yaxis_title = "Частота"
                     , xaxis_title = "Год выпуска видео"
                     , title = {"font": dict(size = 15)
                               , "x": 0.5})
fig231.update_traces(hovertemplate = "Год: %{x} <br>Частота: %{y}")

fig231.show()

Как мы можем видеть на графике существенная популяризация искусственного интелекта начала происходить в последние два года. Что, возможно, в дальнейшем подтвердит нашу гипотезу. Посмотрим как будут отображаться выбросы

In [48]:
fig232 = px.box(data_neiro
                , y = "year"
                , points = "all"
                , template = "simple_white"
                , title = "Ящик с усами для переменной 'Год выпуска видео'")
fig232.update_layout(yaxis_title = "Год выпуска видео"
                     , title = {"font": dict(size = 15)
                                , "x": 0.5})


fig232.show()


In [49]:
# № 23 посмотрим на статистики/показатели для анализа распределения у "Вовлеченности в контент (видео)" и "Года выпуска"
statistic1 = ScaleStatistics(data_neiro, ["involvement", "year"], normality_test = True)


SCALE STATISTICS SUMMARY
------------------




NORMALITY TESTS
------------------



In [51]:
# № 24 посчитаем корреляцию Пирсона
correlation_pearson_1 = Correlation(data_neiro[["involvement"
                                                  , "year"]])


CORRELATION SUMMARY (PEARSON METHOD, 2-TAILED)
------------------


Note: Each empty index duplicates the previous one.
Maximum correlation is 0.03 (p-value 0.376) for variables involvement and year,
minimum correlation is 0.03 (p-value 0.376) for variables involvement and year.


In [52]:
correlation_pearson_1.sort_correlations() # смотрим уникальные коэффициенты корреляции и значимости (p-value)

Note: Each empty index duplicates the previous one.


Coefficient   p-value
involvement year     0.029748  0.376208

In [54]:
fig253 = px.scatter(data_neiro
                    , y = "involvement"
                    , x = "year"
                    , template = "simple_white"
                    , title = "График рассеивания для переменных 'Вовлеченность в контент (видео)' и 'Год выпуска видео'")
fig253.update_layout(yaxis_title = "Вовлеченность в контент (видео), баллы"
                     , xaxis_title = "Год выпуска видео"
                     , title = {"font": dict(size = 15)
                                , "x": 0.5})
fig253.update_traces(hovertemplate = "Балл вовлеченности: %{y} <br>Год: %{x}")

fig253.show()



**Интерпретация результатов, полученных в Задаче 2**


**Интерпретация коэффициента корреляции Пирсона.**
Исходя из рассчитанного коэффициента Пирсона и p-value, можно сделать вывод, что нет статистически значимой связи между вовлеченностью в контент и годом выпуска видео в генеральной совокупности. Таким образом, нулевая гипотеза подтверждается, а альтернативная гипотеза отвергается.


**Интерпретация графика.**
По графику видно, что, чем позже опубликовано видео, тем выше уровень вовлеченности; можно сказать о наличии линейной связи между переменными.

#### Задача 1. Выяснить, связана ли вовлеченность в контент (видео) с категорией видео


**Для поиска парной связи будет использован** критерий Хи-квадрат Пирсона.


**Гипотезы:**
- H0: вовлеченность в контент (видео) не связана с категорией видео, в генеральной совокупности связи между переменными (признаками) нет.
- H1: вовлеченность в контент (видео) связана с категорией видео, в генеральной совокупности связь между переменными (признаками) есть.



In [57]:
# № 25 строим гистограммы и боксплоты для переменных
fig25 = px.histogram(data_neiro
                    , x = "involvement"
                    , template = "simple_white"
                      , title = "Частотное распределение видео по запросу 'нейросеть' ")

fig25.update_layout(yaxis_title = "Частота"
                     , xaxis_title = "Вовлеченность в контент (видео), баллы"
                     , title = {"font": dict(size = 15)
                                , "x": 0.5})

fig25.update_traces(hovertemplate = "Частота: %{y}")

fig25.show()

In [58]:
fig26 = px.box(data_neiro
                , y = "involvement"
                , points = "all"
                , template = "simple_white"
                , title = "Ящик с усами для переменной 'Вовлеченность в контент (видео)'")


fig26.update_layout(yaxis_title = "Вовлеченность в контент (видео), баллы"
                     , title = {"font": dict(size = 15)
                                , "x": 0.5})

fig26.show()

In [63]:
# №26 подготовка данных для построения столбчатой диаграммы для переменной "Категория видео"
data_fig27 = NominalStatistics(data_neiro, ["categoryName"]).frequencies()['categoryName']


NOMINAL STATISTICS SUMMARY
------------------




FREQUENCIES
------------------

variable: categoryName


In [65]:
fig272 = px.bar(data_fig27
                , x = data_fig27.index
                , y = "N"
                , title = "Частотное распределение видео по запросу 'нейросеть' по категориям"
                , template = "simple_white"
                , color = data_fig27.index)

fig272.update_layout(yaxis_title = "Частота"
                     , xaxis_title = ""
                     , title = {"font": dict(size = 15)
                                , "x": 0.5}
                     , showlegend = False)

fig272.update_traces(hovertemplate = "Частота: %{y}")

fig272.show()

In [67]:
# №27 выводим для интервальной переменной "Вовлеченность в контент (видео)" описательные статистики
SS_TASK2 = ScaleStatistics(data_neiro, ["involvement"], normality_test = True)


SCALE STATISTICS SUMMARY
------------------




NORMALITY TESTS
------------------



In [69]:
# №28 выводим для номинальной переменной "Категория видео" описательные статистики
NS_TASK2 = NominalStatistics(data_neiro, ["categoryName"])


NOMINAL STATISTICS SUMMARY
------------------




FREQUENCIES
------------------

variable: categoryName


In [70]:
# №29 выводим список квартилей переменной "Вовлеченность в контент (видео)"
Involvement_Borders = [round(SS_TASK2.summary()["25%"]["involvement"], 2) # 1 квартиль переменной
                       , round(SS_TASK2.summary()["median"]["involvement"], 2) # 2 квартиль или медианное значение по переменной
                       , round(SS_TASK2.summary()["75%"]["involvement"], 2) # 3 квартиль переменной
                       , round(SS_TASK2.summary()["max"]["involvement"], 2)] # Максимальное значение по переменной
Involvement_Borders

[0.04, 0.05, 0.07, 1.33]

In [72]:
# задаем названия групп/категорий для переменной "Вовлеченность в контент (видео)"
Involvement_Labels = ["1 Till 0.04 Low Involvement"
                      , "2 Till 0.05 Lower Medium Involvement"
                      , "3 Till 0.07 Higher Medium Involvement"
                      , "4 Till 1.13 High Involvement"]
Involvement_Labels

['1 Till 0.04 Low Involvement',
 '2 Till 0.05 Lower Medium Involvement',
 '3 Till 0.07 Higher Medium Involvement',
 '4 Till 1.13 High Involvement']

In [73]:
# записываем название для значений группы/категории с низкой вовлеченностью в контент (видео)
data_neiro.loc[data_neiro["involvement"] <= Involvement_Borders[0], "involvement_level"] = Involvement_Labels[0]
data_neiro[data_neiro["involvement"] <= Involvement_Borders[0]] # Выводим видео только с "1 Till 0.01 Low Involvement" и проверяем, все ли правильно записалось

id   snippet.publishedAt         snippet.channelId  \
6    fwR5qlblDzc  2023-05-13T09:43:48Z  UCHoJ7DYrVfKRmD1I8XWjMzA   
8    7XsdO8J1tII  2022-12-08T07:05:15Z  UCkF05U4kMclJJ8Rp6UzrL_w   
9    4eFth-C8O2A  2023-04-20T14:15:00Z  UCVMkXkbfVRmjP615zdvHvLA   
19   0vbk1wG7gqs  2019-11-10T12:13:50Z  UCnlNy4YSuOcs8Qy6kRoMFbw   
20   0HVIIYrBuKg  2022-10-07T20:34:20Z  UChi1KfffybXf30BcqQn6s-Q   
..           ...                   ...                       ...   
882  xaPtYETsIIw  2023-02-20T04:00:30Z  UCrnGjnWe_Q7vl8cR9iRXAfQ   
883  LiSqODp1fTo  2021-08-13T04:51:36Z  UCrlnaTh41nSRX5krPJ2Zqfw   
885  bhsPu9kaMeA  2023-06-04T09:00:20Z  UCSBeL28cCqIyHFxmCTK1Ejw   
886  WGntxEFMdXo  2023-05-12T12:28:57Z  UCOFQ45F6JIVM1Q7KI-gYT3A   
888  WbSSjAo9Y5A  2023-06-13T16:55:36Z  UC0dVruAP10W4T3ixx5doiYA   

                                         snippet.title  \
6    ему больше 30🤣нейросеть, которая определит тво...   
8                     Нейросеть скрестила мем и Dota 2   
9    Президентов нарядили в розовый: новая фантазия...   
19   НА ЧТО СПОСОБЕН ИСКУССТВЕННЫЙ ИНТЕЛЛЕКТ ОТ OPE...   
20   Нейросеть  Предсказала чем закончится контрнас...   
..                                                 ...   
882  Нейросеть изобразила, как выглядят города Росс...   
883  Как найти человека по фото, нейросеть и искусс...   
885             Обучаем Нейросеть без Программирования   
886  Господи, как я смеялась 😂 почему меня так не л...   
888  Нейросеть это будущее которое уже наступило#sh...   

                                   snippet.description  \
6                                                        
8                                                        
9    Ничего милее вы сегодня не увидите. Нейросеть ...   
19   ➥ Подпишись - http://bit.ly/KosmoStory\n➥ Наш ...   
20   В сеть слили планы наступления ВСУ в динамике....   
..                                                 ...   
882  Почти все портреты получились довольно мрачным...   
883  Сейчас XXI век, который по праву считается “Пи...   
885  Обучаем искусственный интеллект без программир...   
886                                                      
888                                                      

                      snippet.channelTitle  snippet.categoryId  \
6                          FINIK & SMOORFY                  22   
8                                amindblow                  23   
9                          Короче, новости                  25   
19                             Kosmo Story                  22   
20                     Хобби и Развлечения                  25   
..                                     ...                 ...   
882                     Интересная планета                  22   
883            LaptopWorld - Мир Ноутбуков                  28   
885  Айтигенио: международная онлайн-школа                  20   
886                            Анна Булгак                  10   
888                           БумНейросети                  22   

    contentDetails.duration  statistics.viewCount  statistics.likeCount  \
6                     PT24S               5949877                230875   
8                     PT18S               5620926                119406   
9                     PT14S               5531638                217689   
19                  PT4M49S               2482997                 94597   
20                  PT2M52S               2352043                 11792   
..                      ...                   ...                   ...   
882                   PT57S                  4313                   143   
883                PT10M26S                  4059                    76   
885                 PT6M23S                  3918                    39   
886                   PT31S                  2667                    63   
888                   PT33S                   198                     5   

     statistics.commentCount  \
6                        283   
8                       1086   
9            

In [74]:
# записываем названия для значений остальных групп/категорий переменной "Уровень вовлеченности в контент (видео)" с помощью цикла for
for i in range(1, len(Involvement_Borders)):
    data_neiro.loc[(data_neiro["involvement"] > Involvement_Borders[i-1])
                  & (data_neiro["involvement"] <= Involvement_Borders[i])
                 , "involvement_level"] = Involvement_Labels[i] # задаем условие для записи оставшихся названий групп/категорий
    print(f"Итерация {i}")

data_neiro

Итерация 1
Итерация 2
Итерация 3


id   snippet.publishedAt         snippet.channelId  \
1    3OxtnK1m5Sw  2022-10-20T22:44:44Z  UCm7bpbe7_H_47ef4dJH0SNw   
2    h6QZtd-Efko  2023-05-04T09:59:13Z  UCf2ZCvUQRwdkFUfebyy0Xag   
3    HaLZtHN83JE  2023-01-26T13:00:15Z  UCRd8_F4-SynaG4sE5lnXv_g   
4    gkDrVYInSaQ  2023-02-10T14:00:24Z  UCUEQBzSZx65-FcVwce8HYNQ   
5    P8GXJ7AWqIA  2022-12-05T07:04:35Z  UCkF05U4kMclJJ8Rp6UzrL_w   
..           ...                   ...                       ...   
884  ZxIN6vCIMe8  2022-06-19T12:36:11Z  UCxMBGX-c9pGG-YOGzdMV1pQ   
885  bhsPu9kaMeA  2023-06-04T09:00:20Z  UCSBeL28cCqIyHFxmCTK1Ejw   
886  WGntxEFMdXo  2023-05-12T12:28:57Z  UCOFQ45F6JIVM1Q7KI-gYT3A   
887  j3od_Wn58DQ  2023-04-25T21:10:31Z  UC-zQQEt2z8kO-lyhaJIJmaQ   
888  WbSSjAo9Y5A  2023-06-13T16:55:36Z  UC0dVruAP10W4T3ixx5doiYA   

                                         snippet.title  \
1                     Топ-3 нейросети #айти #нейросети   
2                 Лютая сладость от нейросети) Часть 1   
3    Как создать игру если Искусственный Интеллект ...   
4    Нейросеть побила мировой рекорд! #игры #компью...   
5       Нейросеть превращает изображения в АНИМЕ Vol.2   
..                                                 ...   
884                  Нейросеть пояснила за Prey (2006)   
885             Обучаем Нейросеть без Программирования   
886  Господи, как я смеялась 😂 почему меня так не л...   
887  Новая бесплатная нейросеть Kandinsky 2.1 от Сб...   
888  Нейросеть это будущее которое уже наступило#sh...   

                                   snippet.description  \
1                                                        
2                                                        
3    Этот туториал поможет вам ответить на вопрос, ...   
4                                                        
5                                                        
..                                                 ...   
884  В сегодняшней телепрограмме мы узнаем, что же ...   
885  Обучаем искусственный интеллект без программир...   
886                                                      
887  Кандинский — обновленная нейросеть, которая ге...   
888                                                      

                      snippet.channelTitle  snippet.categoryId  \
1                              YCLA-Coding                  22   
2                            Добрый Джонни                  22   
3                             GameDev Club                  24   
4                  Роман Сакутин - GameDev                  27   
5                                amindblow                  22   
..                                     ...                 ...   
884                            Qwetux 2009                   1   
885  Айтигенио: международная онлайн-школа                  20   
886                            Анна Булгак                  10   
887               Нейросетевая Лаборатория                  28   
888                           БумНейросети                  22   

    contentDetails.duration  statistics.viewCount  statistics.likeCount  \
1                     PT36S              11005576                801306   
2                     PT59S              10420652                717458   
3                     PT54S               9157840                420422   
4                     PT42S               6481887                320103   
5                     PT21S               6269361                293735   
..                      ...                   ...                   ...   
884                 PT4M33S                  3544                   328   
885                 PT6M23S                  3918                    39   
886                   PT31S                  2667                    63   
887                 PT8M24S                  2139                   104   
888                   PT33S                   198                     5   

     statistics.commentCount  \
1                       2094   
2                       1137   
3            

In [75]:
# №30 вывели таблицу сопреженности
CROSSTAB_TASK2 = Crosstab(data_neiro, "categoryName", "involvement_level")


CROSSTAB SUMMARY
------------------

Observed frequencies


------------------

Expected frequencies


25 (41.667%) cells have expected frequency less than 5. The minimum expected frequency is 0.152.
------------------

Chi-square statistic is 84.264 (p-value = 0.0).
------------------

Pearson's residuals


5 (8.333%) cells have Pearson's residual bigger than 1.96. 
The biggest residual is 5.078 (categories Gaming and 4 Till 1.13 High Involvement).
The smallest residual is -3.49 (categories Gaming and 1 Till 0.04 Low Involvement).


In [76]:
CROSSTAB_TASK2.pvalue #смотрим p-value

0.00011828824913885281

In [77]:
# выводим отдельно положительные значимые остатки критерия Хи-квадрат
CROSSTAB_TASK2.residuals_pearson > 1.96

involvement_level      1 Till 0.04 Low Involvement  \
categoryName                                         
Autos & Vehicles                             False   
Comedy                                       False   
Education                                    False   
Entertainment                                False   
Film & Animation                             False   
Gaming                                       False   
Howto & Style                                False   
Music                                        False   
News & Politics                              False   
Nonprofits & Activism                        False   
People & Blogs                                True   
Pets & Animals                               False   
Science & Technology                         False   
Sports                                       False   
Travel & Events                              False   

involvement_level      2 Till 0.05 Lower Medium Involvement  \
categoryName                                                  
Autos & Vehicles                                      False   
Comedy                                                False   
Education                                             False   
Entertainment                                         False   
Film & Animation                                      False   
Gaming                                                False   
Howto & Style                                         False   
Music                                                 False   
News & Politics                                        True   
Nonprofits & Activism                                 False   
People & Blogs                                        False   
Pets & Animals                                        False   
Science & Technology                                  False   
Sports                                                False   
Travel & Events                                       False   

involvement_level      3 Till 0.07 Higher Medium Involvement  \
categoryName                                                   
Autos & Vehicles                                       False   
Comedy                                                 False   
Education                                              False   
Entertainment                                          False   
Film & Animation                                       False   
Gaming                                                 False   
Howto & Style                                          False   
Music                                                  False   
News & Politics                                        False   
Nonprofits & Activism                                  False   
People & Blogs                                         False   
Pets & Animals                                         False   
Science & Technology                                   False   
Sports                                                 False   
Travel & Events                                        False   

involvement_level      4 Till 1.13 High Involvement  
categoryName                                         
Autos & Vehicles                              False  
Comedy                                        False  
Education                                     False  
Entertainment                                 False  
Film & Animation                              False  
Gaming                                         True  
Howto & Style                                 False  
Music                                         False  
News & Politics                               False  
Nonprofits & Activism                         False  
People & Blogs                                False  
Pets & Animals                                False  
Science & Technology                          False  
Sports                                        False  
Travel & Events                               False

In [78]:
# тоже самое для отдельно отрицательных значимых остатков критерия Хи-квадрат
CROSSTAB_TASK2.residuals_pearson < -1.96

involvement_level      1 Till 0.04 Low Involvement  \
categoryName                                         
Autos & Vehicles                             False   
Comedy                                       False   
Education                                    False   
Entertainment                                False   
Film & Animation                             False   
Gaming                                        True   
Howto & Style                                False   
Music                                        False   
News & Politics                              False   
Nonprofits & Activism                        False   
People & Blogs                               False   
Pets & Animals                               False   
Science & Technology                         False   
Sports                                       False   
Travel & Events                              False   

involvement_level      2 Till 0.05 Lower Medium Involvement  \
categoryName                                                  
Autos & Vehicles                                      False   
Comedy                                                False   
Education                                             False   
Entertainment                                         False   
Film & Animation                                      False   
Gaming                                                False   
Howto & Style                                         False   
Music                                                 False   
News & Politics                                       False   
Nonprofits & Activism                                 False   
People & Blogs                                        False   
Pets & Animals                                        False   
Science & Technology                                  False   
Sports                                                False   
Travel & Events                                       False   

involvement_level      3 Till 0.07 Higher Medium Involvement  \
categoryName                                                   
Autos & Vehicles                                       False   
Comedy                                                 False   
Education                                              False   
Entertainment                                          False   
Film & Animation                                       False   
Gaming                                                 False   
Howto & Style                                          False   
Music                                                  False   
News & Politics                                        False   
Nonprofits & Activism                                  False   
People & Blogs                                         False   
Pets & Animals                                         False   
Science & Technology                                   False   
Sports                                                 False   
Travel & Events                                        False   

involvement_level      4 Till 1.13 High Involvement  
categoryName                                         
Autos & Vehicles                              False  
Comedy                                        False  
Education                                     False  
Entertainment                                 False  
Film & Animation                              False  
Gaming                                        False  
Howto & Style                                 False  
Music                                         False  
News & Politics                               False  
Nonprofits & Activism                         False  
People & Blogs                                 True  
Pets & Animals                                False  
Science & Technology                          False  
Sports                                        False  
Travel & Events                               False

In [79]:
# сохраняем чтоб наверняка наши данные в excel
data_neiro.to_excel("Проект нейросеть.xlsx")


**Интерпретация результатов для критерия Хи-квадрат.**
На уровне 95% доверительной вероятности нулевая гипотеза об отсутствии связи отклоняется (p-value = 0.0, что меньше, чем 0.05). Следовательно, связь между вовлеченностью в контент (видео) и категорией видео есть. Однако, поскольку процент ненаполненных ячеек превышает допустимое пороговое значение (20%) и составляет 84.264%, доверять полученным результатам критерия Хи-квадрат нельзя. В таком случае мы можем проанализировать и проинтерпретировать статистически значимые остатки критерия Хи-квадрат.

Исходя из рассчитанного значения Хи-квадрат и p-value, можно сделать вывод, что есть статистически значимая связь между вовлеченностью в контент и категорией видео в генеральной совокупности. Таким образом, нулевая гипотеза отвергается, а альтернативная гипотеза подтверждается. Значение Хи-квадрат показывает, что различия между ожидаемыми и фактическими значениями частот значительны. Результаты Pearson's residual показывают, какие ячейки таблицы имеют наибольшее отклонение от ожидаемых значений. Например, категории Gaming и высокая вовлеченность имеют наибольшее положительное отклонение. С другой стороны, категории People & Blogs и низкая вовлеченность имеют наибольшее отрицательное отклонение, что может указывать на то, что видео в этой категории чаще вызывают высокую вовлеченность зрителей. В целом, результаты анализа позволяют сделать вывод о наличии связи между вовлеченностью в контент и категорией видео в генеральной совокупности.